File: c/examples/tcp_relay.c
/* SPDX-FileCopyrightText: 2026 Alexander R. Croft */
/* SPDX-License-Identifier: MIT */

#include "ratatouille.h"

#include <stdio.h>

int main(void) {
    rat_tcp_relay_config_t relay_cfg = {0};
    rat_config_t cfg = {0};
    rat_tcp_relay_t *relay;
    rat_logger_t *log;
    rat_tcp_relay_stats_t relay_stats;

    relay_cfg.endpoint = "tcp://127.0.0.1:9000";
    relay_cfg.batch_bytes = 4096;
    relay_cfg.max_queue_bytes = 65536;
    relay_cfg.max_queue = 128;
    relay_cfg.drop_policy = RAT_DROP_OLDEST;

    relay = rat_tcp_relay_create(&relay_cfg);
    if (!relay) {
        fprintf(stderr, "failed to create TCP relay\n");
        return 1;
    }

    cfg.filter = "api*";
    cfg.format = RAT_FORMAT_NDJSON;
    cfg.sink = rat_tcp_relay_callback;
    cfg.sink_userdata = relay;
    cfg.source.app = "example";
    cfg.source.where = "c";
    cfg.source.instance = "local";

    log = rat_logger_create(&cfg);
    if (!log) {
        fprintf(stderr, "failed to create logger\n");
        rat_tcp_relay_destroy(relay);
        return 1;
    }

    rat_log(log, "api", "queued one");
    rat_log(log, "api", "queued two");
    rat_logf(log, "api", "queued value=%d", 3);

    if (rat_tcp_relay_flush_now(relay) < 0) {
        fprintf(stderr, "relay flush failed\n");
    }

    relay_stats = rat_tcp_relay_stats(relay);
    fprintf(
        stderr,
        "queued=%llu dropped=%llu sent_batches=%llu sent_bytes=%llu failed_flushes=%llu\n",
        (unsigned long long)relay_stats.queued,
        (unsigned long long)relay_stats.dropped,
        (unsigned long long)relay_stats.sent_batches,
        (unsigned long long)relay_stats.sent_bytes,
        (unsigned long long)relay_stats.failed_flushes
    );

    rat_logger_destroy(log);
    rat_tcp_relay_destroy(relay);
    return 0;
}

File: rust/examples/tcp_relay.rs
// SPDX-FileCopyrightText: 2026 Alexander R. Croft
// SPDX-License-Identifier: MIT

use ratatouille::{Format, Logger, LoggerConfig, SourceIdentity, TcpRelay, TcpRelayConfig};

fn main() -> std::io::Result<()> {
    let relay = TcpRelay::new(TcpRelayConfig {
        endpoint: "tcp://127.0.0.1:9000".into(),
        batch_bytes: 4096,
        max_queue_bytes: 65_536,
        max_queue: 128,
        ..TcpRelayConfig::default()
    })?;

    let mut config = LoggerConfig::default();
    config.filter = Some("api*".into());
    config.format = Format::Ndjson;
    config.source = SourceIdentity {
        app: Some("example".into()),
        r#where: Some("rust".into()),
        instance: Some("local".into()),
    };

    let mut log = Logger::with_sink(config, relay);
    let _ = log.log("api", "queued one");
    let _ = log.log("api", "queued two");
    let _ = log.log("api", "queued value=3");
    let _ = log.sink_mut().flush_now()?;

    let stats = log.sink_mut().stats();
    eprintln!(
        "queued={} dropped={} sent_batches={} sent_bytes={} failed_flushes={}",
        stats.queued, stats.dropped, stats.sent_batches, stats.sent_bytes, stats.failed_flushes
    );
    Ok(())
}